pkg/storage/permanent/settings.go

```diff
@@ -1,264 +1,440 @@
+func (s *Settings) StoreProtocolParametersEpochMapping(version iotago.Version, epoch iotago.EpochIndex) error {
 	s.mutex.Lock()
 	defer s.mutex.Unlock()
 
-	s.settingsModel.ProtocolParameters = params
-	s.UpdateAPI()
+	s.apiMutex.Lock()
+	defer s.apiMutex.Unlock()
 
-	if err = s.ToFile(); err != nil {
-		return errors.Wrap(err, "failed to persist initialized flag")
+	bytes, err := epoch.Bytes()
+	if err != nil {
+		return err
 	}
 
+	if err := s.store.Set([]byte{protocolVersionEpochMappingKey, version}, bytes); err != nil {
+		return err
+	}
+
+	s.protocolVersions = append(s.protocolVersions, &protocolVersionEpochStart{
+		Version:    version,
+		StartEpoch: epoch,
+	})
+
 	return nil
 }
 
 func (s *Settings) LatestCommitment() *model.Commitment {
 	s.mutex.RLock()
-	if s.latestCommitment == nil {
-		s.mutex.RUnlock()
-		s.mutex.Lock()
-		defer s.mutex.Unlock()
+	bytes, err := s.store.Get([]byte{latestCommitmentKey})
+	s.mutex.RUnlock() // do not use defer because s.LatestAPI will also RLock the mutex
 
-		if s.api.TimeProvider().SlotDurationSeconds() == 0 {
-			panic("accessing the LatestCommitment before the settings are initialized")
+	if err != nil {
+		if ierrors.Is(err, kvstore.ErrKeyNotFound) {
+			return model.NewEmptyCommitment(s.LatestAPI())
 		}
-		s.latestCommitment = lo.PanicOnErr(model.CommitmentFromCommitment(s.settingsModel.LatestCommitment, s.api))
-
-		return s.latestCommitment
+		panic(err)
 	}
-	defer s.mutex.RUnlock()
 
-	return s.latestCommitment
+	return lo.PanicOnErr(model.CommitmentFromBytes(bytes, s))
 }
 
 func (s *Settings) SetLatestCommitment(latestCommitment *model.Commitment) (err error) {
 	s.mutex.Lock()
 	defer s.mutex.Unlock()
 
-	s.settingsModel.LatestCommitment = latestCommitment.Commitment()
-
-	if err = s.ToFile(); err != nil {
-		return errors.Wrap(err, "failed to persist latest commitment")
-	}
-
-	s.latestCommitment = latestCommitment
-
-	return nil
+	return s.store.Set([]byte{latestCommitmentKey}, latestCommitment.Data())
 }
 
 func (s *Settings) LatestFinalizedSlot() iotago.SlotIndex {
 	s.mutex.RLock()
 	defer s.mutex.RUnlock()
 
-	return s.settingsModel.LatestFinalizedSlot
+	bytes, err := s.store.Get([]byte{latestFinalizedSlotKey})
+	if err != nil {
+		if ierrors.Is(err, kvstore.ErrKeyNotFound) {
+			return 0
+		}
+		panic(err)
+	}
+	i, _, err := iotago.SlotIndexFromBytes(bytes)
+	if err != nil {
+		panic(err)
+	}
+
+	return i
 }
 
 func (s *Settings) SetLatestFinalizedSlot(index iotago.SlotIndex) (err error) {
 	s.mutex.Lock()
 	defer s.mutex.Unlock()
 
-	s.settingsModel.LatestFinalizedSlot = index
+	return s.store.Set([]byte{latestFinalizedSlotKey}, index.MustBytes())
+}
 
-	if err = s.ToFile(); err != nil {
-		return errors.Wrap(err, "failed to persist latest confirmed slot")
+func (s *Settings) ProtocolParametersAndVersionsHash() ([32]byte, error) {
+	// TODO: concatenate and hash protocol parameters and protocol versions
```

Comment:
Was this left out intentionally?

Reply:
Yes. This will be done in a different PR.